classes/helper.php

```diff
@@ -131,4 +131,50 @@ public static function enable_notifications() : string {
         });
         </script>" . PHP_EOL;
     }
+
+    /**
+     * Merge duplicate local_csp records that have matching sha1hash fields.
+     *
+     * @param array $records Array of local_csp records with matching sha1hash. Must contain at least two records.
+     * @return \stdClass The final object representing a local_csp record once all records are merged into one.
+     */
+    public static function merge_duplicate_records(array $records): \stdClass {
+        global $DB;
+        // If no records provided, there is no object to return so throw exception.
+        if (empty($records)) {
+            throw new \moodle_exception('norecordsfound', 'local_csp');
+        }
+
+        // If only 1 record provided, just return it.
+        if (count($records) === 1) {
+            return reset($records);
+        }
+
+        // Order records by DESC failcounter, as highest fail counter will be the base record.
+        usort($records, function($a, $b) {
```

Comment:
Does this actually matter? they should be identical. The only thing I can think of is timestamps, otherwise we can be a little faster if the sort isnt needed

Reply:
Yeah timestamps were why I did (and being unfamiliar with significance of different data). If you don't think which record ultimately is kept then I can remove this.

Reply:
One thing that becomes really annoying is testing if it's not sorted, though.